code-block: Fix handling of :dedent: and add unit tests

latosha-maltba opened the pull request:
Subject: code-block: Fix handling of :dedent: and add unit tests

### Feature or Bugfix
Depending on the point of view this is one or a combination of:
- Feature
- Bugfix
- Refactoring

### Purpose
- an argument of ``0`` to dedent is handled as numeric argument and not as "no argument".
- Preserve newlines in the literal block when :dedent: without argument is given.

### Relates
- #9636

Commit:
Add tests for :dedent: option of code-block directive

Commit:
code-block: Handle ``0`` as numeric argument in :dedent:

Due to how Python converts ints to bool the numeric argument ``0`` to
the :dedent: option of the code-block directive was handled like if no
argument was given to :dedent:.

Check properly for None in the options argument processing, so ``0`` is
treated as numeric argument to :dedent:.

Bug: #9636

Commit:
Fix: 9636: code-block: Do not remove newlines with :dedent:

If the option :dedent: (without numeric argument) was given to
code-block then all newlines where striped from the resulting code.
Preserve newlines when stripping common whitespace at the beginning of
code blocks.

tk0miya commented on doc/usage/restructuredtext/directives.rst:
```diff
@@ -598,10 +598,23 @@ __ https://pygments.org/docs/lexers
       are removed via :func:`textwrap.dedent()`.  For example::
 
          .. code-block:: ruby
+            :linenos:
```
Why did you add this option here?

tk0miya replied:
Hmm? This example contains both `:linenos:` and `:dedent: 4` option. I don't still understand why you added the former one.

>(it mostly affects code as whitespace is usually not significant in prose)

I'm not sure because there are many kinds of directives (including 3rd parties) in the world. So there might be some directives that give a special meaning to indentation. IMO, this is common problem (or tip) for the reST directive, not only for `code-block`.

tk0miya replied:
Thank you for your explanation. I understood what you'd like to do.

In my thought, this paragraph is a description of the `:dedent:` option. So it should describe how it works. But it seems you're going to describe what is happened if the option not given, right?

Indeed, the examples you shown are converted to the same result. But, the meaning and its internal processing are different at all. So they should be separated.

>It is and shouldn't it be therefore mentioned more prominently?
>It does not seem to be documented anywhere except buried deep in some specification.

IMO, the best place is the spec of reST. Our document does not describe the whole of the syntax. So it's best.
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#directives

Another idea is the "Directive" section of the "reST Primer" document. It describes the indentation of the directives as following. It would be better to make it details.

>The directive content follows after a blank line and is indented relative to the directive start.

tk0miya commented on doc/usage/restructuredtext/directives.rst:
```diff
@@ -598,10 +598,23 @@ __ https://pygments.org/docs/lexers
 
                 some ruby code
 
+      Using ``0`` as argument does not strip anything and can be used to
```
As I commented in the issue, I can agree on this feature as a "secret" feature. So this should not be mentioned in the document. To add a proper feature, we have to consider another way to keep indentation because it's common problem of reST directives, not only this directive.

Commit:
Update :dedent: example

The presence or absence of the ``:dedent: 4`` option in the example

.. code-block:: ruby
   :dedent: 4

       some ruby code

makes no difference for the output due to how ReST handles whitespace.
If it is present, then ReST indent is 3 and the 4 spaces of the code are
stripped by the :dedent: option.  If the dedent option is missing, then
the ReST indent is 7 and all whitespace is stripped again.

Add a dummy option to fixate the ReST indent to 3, so adding and
removing of the option in the example has an effect on the output.

tk0miya committed:
Merge branch '4.x' into HEAD

tk0miya committed:
doc: Remove a mention to :dedent: 0

`:dedent: 0` option for the code-block directive is a secret feature.
So it should not be mentioned in our docs.

tk0miya committed:
Fix a flake8 warning

tk0miya committed:
code-block: Pass list of strings ends with CR code to dedent_lines()
